csharp/src/Apache.Arrow/C/CArrowSchema.cs

```diff
@@ -0,0 +1,519 @@
+        public IntPtr private_data;
+
+        private static string GetFormat(IArrowType datatype)
+        {
+            TypeFormatter formatter = new TypeFormatter();
+            datatype.Accept(formatter);
+            return formatter.format_string;
+        }
+
+        private static long GetFlags(IArrowType datatype, bool nullable = true)
+        {
+            long flags = 0;
+
+            if (nullable)
+            {
+                flags |= ARROW_FLAG_NULLABLE;
+            }
+
+            if (datatype is DictionaryType)
+            {
+                if (((DictionaryType)datatype).Ordered)
+                {
+                    flags |= ARROW_FLAG_DICTIONARY_ORDERED;
+                }
+            }
+
+            // TODO: when we implement MapType, make sure to set the KEYS_SORTED flag.
+            return flags;
+        }
+
+        private static IntPtr* ConstructChildren(IArrowType datatype)
+        {
+            if (datatype is NestedType)
+            {
+                var fields = ((NestedType)datatype).Fields;
+                int n_fields = fields.Count;
+
+                IntPtr* pointer_list = (IntPtr*)Marshal.AllocHGlobal(n_fields * sizeof(IntPtr));
+
+                for (var i = 0; i < n_fields; i++)
+                {
+                    var c_schema = new CArrowSchema();
+                    CArrowSchema.ExportField(fields[i], out c_schema);
+                    IntPtr exported_schema = c_schema.AllocateAsPtr();
+                    pointer_list[i] = exported_schema;
+                }
+
+                return pointer_list;
+            }
+            else
+            {
+                return (IntPtr*)IntPtr.Zero;
+            }
+        }
+
+        private static IntPtr ConstructDictionary(IArrowType datatype)
+        {
+            if (datatype is DictionaryType)
+            {
+                var c_schema = new CArrowSchema();
+                var value_type = ((DictionaryType)datatype).ValueType;
+                CArrowSchema.ExportDataType(value_type, out c_schema);
+                return c_schema.AllocateAsPtr();
+            }
+            else
+            {
+                return IntPtr.Zero;
+            }
+        }
+
+        /// <summary>
+        /// Initialize the exported C schema as an Arrow type.
+        /// </summary>
+        /// <param name="datatype">The Arrow type to export.</param>
+        /// <param name="schema">An uninitialized CArrowSchema.</param>
+        public static void ExportDataType(IArrowType datatype, out CArrowSchema schema)
+        {
+            schema.format = GetFormat(datatype);
+            schema.name = null;
+            schema.metadata = null;
+            schema.flags = GetFlags(datatype);
+
+            schema.children = ConstructChildren(datatype);
+            schema.n_children = datatype is NestedType ? ((NestedType)datatype).Fields.Count : 0;
```

Comment:
How about setting them at once in `ConstructChildren()` because they are related?

Reply:
I was originally trying to see if I could do something like

```csharp
(schema.children, schema.n_children) = ConstructChildren(datatype);
```

but couldn't figure out how to import support for tuples. I also tried a signature like:

```csharp
void SetChildren(datatype, out CArrowSchema schema);
```

But then C# wants me to set every field of `schema` within `SetChildren` (it doesn't know it's only being called from a function that does).

Open to suggestions :)